File: Codes/144_preorderTraversal/144.二叉树的前序遍历.go
/*
 * @Description:
 * @Autor: HTmonster
 * @Date: 2022-02-16 20:09:38
 */
/*
 * @lc app=leetcode.cn id=144 lang=golang
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// 递归法
// func preorderTraversal(root *TreeNode) (result []int) {
// 	var dfs func(*TreeNode)

// 	dfs = func(node *TreeNode) { //递归函数
// 		if node == nil {
// 			return //退出条件
// 		}
// 		result = append(result, node.Val) //1. 记录本节点
// 		dfs(node.Left)                    //2. 遍历左节点
// 		dfs(node.Right)                   //3. 遍历右节点
// 	}

// 	dfs(root)
// 	return
// }
func preorderTraversal(root *TreeNode) (result []int) {
	stack := make([]*TreeNode, 0)

	if root == nil {
		return
	}
	stack = append(stack, root) // 入根节点

	for len(stack) > 0 {
		node := stack[len(stack)-1] //1. 弹出
		stack = stack[:len(stack)-1]

		result = append(result, node.Val) //2. 记录
		if node.Right != nil {
			stack = append(stack, node.Right) //3.入栈右子树
		}
		if node.Left != nil {
			stack = append(stack, node.Left) //4.入栈左子树
		}
	}
	return
}

// @lc code=end

File: Codes/144_preorderTraversal/144.二叉树的前序遍历.py
'''
Description: 
Autor: HTmonster
Date: 2022-02-16 19:55:34
'''
#
# @lc app=leetcode.cn id=144 lang=python3
#
# [144] 二叉树的前序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # 递归法
    # def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
    #     result=[]
    #     def dfs(node):
    #         if not node: return       # 退出条件
    #         nonlocal result
    #         result.append(node.val)   # 1.保存根节点

    #         dfs(node.left)            # 2.遍历左节点
    #         dfs(node.right)           # 3.遍历右节点

    #     dfs(root)
    #     return result

    # 迭代法
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        stack,result=[],[]
        
        if root: stack.append(root)             # 根节点首先入站
        while len(stack) > 0:
            node=stack.pop()                    # 1弹出第一个
            result.append(node.val)             # 添加值

            if node.right:
                stack.append(node.right)        # 2加右子树
            
            if node.left:
                stack.append(node.left)         # 3加左子树

        return result
# @lc code=end

